rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp

```diff
@@ -0,0 +1,401 @@
+
+      srv_get_state_ = std::make_shared<rclcpp::service::Service<GetStateSrv>>(
+        node_base_interface_->get_shared_rcl_node_handle(),
+        &state_machine_.com_interface.srv_get_state,
+        any_cb);
+      node_services_interface_->add_service(
+        std::dynamic_pointer_cast<rclcpp::service::ServiceBase>(srv_get_state_),
+        nullptr);
+    }
+
+    {  // get_available_states
+      auto cb = std::bind(&LifecycleNodeInterfaceImpl::on_get_available_states, this,
+          std::placeholders::_1, std::placeholders::_2, std::placeholders::_3);
+      rclcpp::any_service_callback::AnyServiceCallback<GetAvailableStatesSrv> any_cb;
+      any_cb.set(std::move(cb));
+
+      srv_get_available_states_ = std::make_shared<rclcpp::service::Service<GetAvailableStatesSrv>>(
+        node_base_interface_->get_shared_rcl_node_handle(),
+        &state_machine_.com_interface.srv_get_available_states,
+        any_cb);
+      node_services_interface_->add_service(
+        std::dynamic_pointer_cast<rclcpp::service::ServiceBase>(srv_get_available_states_),
+        nullptr);
+    }
+
+    {  // get_available_transitions
+      auto cb = std::bind(&LifecycleNodeInterfaceImpl::on_get_available_transitions, this,
+          std::placeholders::_1, std::placeholders::_2, std::placeholders::_3);
+      rclcpp::any_service_callback::AnyServiceCallback<GetAvailableTransitionsSrv> any_cb;
+      any_cb.set(std::move(cb));
+
+      srv_get_available_transitions_ =
+        std::make_shared<rclcpp::service::Service<GetAvailableTransitionsSrv>>(
+        node_base_interface_->get_shared_rcl_node_handle(),
+        &state_machine_.com_interface.srv_get_available_transitions,
+        any_cb);
+      node_services_interface_->add_service(
+        std::dynamic_pointer_cast<rclcpp::service::ServiceBase>(srv_get_available_transitions_),
+        nullptr);
+    }
+  }
+
+  bool
+  register_callback(std::uint8_t lifecycle_transition, std::function<rcl_lifecycle_ret_t(
+      const State &)> & cb)
+  {
+    cb_map_[lifecycle_transition] = cb;
+    return true;
+  }
+
+  void
+  on_change_state(const std::shared_ptr<rmw_request_id_t> header,
+    const std::shared_ptr<ChangeStateSrv::Request> req,
+    std::shared_ptr<ChangeStateSrv::Response> resp)
+  {
+    (void)header;
+    if (rcl_lifecycle_state_machine_is_initialized(&state_machine_) != RCL_RET_OK) {
+      resp->success = false;
+      return;
+    }
+    resp->success = change_state(req->transition.id);
+  }
+
+  void
+  on_get_state(const std::shared_ptr<rmw_request_id_t> header,
+    const std::shared_ptr<GetStateSrv::Request> req,
+    std::shared_ptr<GetStateSrv::Response> resp)
+  {
+    (void)header;
+    (void)req;
+    if (rcl_lifecycle_state_machine_is_initialized(&state_machine_) != RCL_RET_OK) {
+      resp->current_state.id =
+        static_cast<uint8_t>(lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN);
+      resp->current_state.label = "unknown";
+      return;
```

Comment:
Since the state machine is initialized before the service, within the same function, is ever possible? Maybe instead of having an unknown state, this should just raise an exception if somehow it is called before the state machine is initialized.